src/afa/afa.cc

```diff
@@ -34,29 +36,531 @@ const std::string Mata::Afa::TYPE_AFA = "AFA";
+{
+    ClosedSet<State> result(ClosedSet<State>::downward_closed, 0, transRelation.size()-1);
+    for(auto node : nodes)
+    {
+        result = result.Union(pre(node, symb));
+    }
+    return result;
+} // pre }}}
+
+/** This function performs pre directly over the given closed set and the given symbol
+* The result is always a downward closed set!
+* @param closed_set a closed set
+* @param symb a symbol used to perform an inverse transition
+* @return closed set of nodes
+*/
+ClosedSet<State> Afa::pre(ClosedSet<State> closed_set, Symbol symb) const
+{
+    assert(closed_set.get_type() == ClosedSet<State>::downward_closed);
+    return pre(closed_set.get_antichain(), symb);
+} // pre }}}
+
+/** This function allows us to perfom pre over a node and use the whole alphabet to perform individual transitions 
+* The result is always a downward closed set!
+* @param node a set of states
+* @return closed set of nodes
+*/
+ClosedSet<State> Afa::pre(Node node) const
+{
+    if(node.empty())
+    {
+        return ClosedSet<State>(ClosedSet<State>::downward_closed, 0, transRelation.size()-1, Nodes{Node{}});
+    }
+    ClosedSet<State> result(ClosedSet<State>::downward_closed, 0, transRelation.size()-1);
+    for(auto transVec : invTransRelation[*(node.begin())])
+    {
+        result.insert(pre(node, transVec.symb).get_antichain());
+    }
+    return result;
+} // pre }}}
+
+/** This function allows us to perfom pre over a set of nodes and use the whole alphabet to perform individual transitions 
+* The result is always a downward closed set!
+* @param nodes a set of sets of states
+* @return closed set of nodes
+*/
+ClosedSet<State> Afa::pre(Nodes nodes) const
+{
+    ClosedSet<State> result(ClosedSet<State>::downward_closed, 0, transRelation.size()-1);
+    for(auto node : nodes)
+    {
+        result.insert(pre(node).get_antichain());
+    }
+    return result;
+} // pre }}}
+
+// The end of the definitions of the "pre" functions
+////////////////////////////////////////////////////
+
 bool Afa::has_trans(const Trans& trans) const
 { // {{{
-  return std::find(this->transitions.begin(),
-                   this->transitions.end(),
-                   trans) != this->transitions.end();
+    Nodes res = *perform_trans(trans.src, trans.symb);
+    if(res.size() && res.IsSubsetOf(trans.dst))
+    {
+        return true;
+    }
+    return false;
 } // has_trans }}}
 
 
 size_t Afa::trans_size() const
 { // {{{
-	return this->transitions.size();
+    size_t result = 0;
+    for(auto state : transRelation)
+    {
+        result += state.size();
+    }
+    return result;
 } // trans_size() }}}
 
+/** This function takes all the initial states and creates a corresponding upward-closed set
+* @return closed set of initial nodes
+*/
+ClosedSet<State> Afa::get_initial_nodes(void) const
+{
+    ClosedSet<State> result(ClosedSet<State>::upward_closed, 0, transRelation.size()-1);
+    for(State state = 0; state < transRelation.size(); ++state)
```

Comment:
Thanks for your suggestion, I have reimplemented it.
It was possible to easily change the function `get_final_nodes()` since the corresponding antichain has a format `{{f_1, f_2, f_3, ... f_n}}`, where `f_1, f_2, ... f_n` are final states. However, the function `get_initial_nodes()` needs to create an antichain which has a format `{{s_1}, {s_2}, {s_3}, ... {s_n}}`, where `s_1, s_2, ... s_n` are initial states so I need to create several nodes and I cannot easily use the `initialstates` set.